--- cpp/neuralnet/modelversion.cpp [lightvector:master]+++ cpp/neuralnet/modelversion.cpp [hzyhhzy:Reversi2022]@@ -42,15 +42,7 @@ }
 
 int NNModelVersion::getNumSpatialFeatures(int modelVersion) {
-  if(modelVersion == 3 || modelVersion == 4)
-    return NNInputs::NUM_FEATURES_SPATIAL_V3;
-  else if(modelVersion == 5)
-    return NNInputs::NUM_FEATURES_SPATIAL_V4;
-  else if(modelVersion == 6)
-    return NNInputs::NUM_FEATURES_SPATIAL_V5;
-  else if(modelVersion == 7)
-    return NNInputs::NUM_FEATURES_SPATIAL_V6;
-  else if(modelVersion == 8 || modelVersion == 9 || modelVersion == 10)
+  if(modelVersion == 8 || modelVersion == 9 || modelVersion == 10)
     return NNInputs::NUM_FEATURES_SPATIAL_V7;
 
   fail(modelVersion);
@@ -58,15 +50,7 @@ }
 
 int NNModelVersion::getNumGlobalFeatures(int modelVersion) {
-  if(modelVersion == 3 || modelVersion == 4)
-    return NNInputs::NUM_FEATURES_GLOBAL_V3;
-  else if(modelVersion == 5)
-    return NNInputs::NUM_FEATURES_GLOBAL_V4;
-  else if(modelVersion == 6)
-    return NNInputs::NUM_FEATURES_GLOBAL_V5;
-  else if(modelVersion == 7)
-    return NNInputs::NUM_FEATURES_GLOBAL_V6;
-  else if(modelVersion == 8 || modelVersion == 9 || modelVersion == 10)
+  if(modelVersion == 8 || modelVersion == 9 || modelVersion == 10)
     return NNInputs::NUM_FEATURES_GLOBAL_V7;
 
   fail(modelVersion);
